spec.bs

```diff
@@ -44,19 +58,89 @@ This specification extends the client-side storage available beyond cookies.
+
+<pre class="idl">
+dictionary StorageAccessTypes {
+  boolean all = false;
+  boolean cookies = false;
+};
+
+interface StorageAccessHandle {
+};
+
+partial interface Document {
+  Promise&lt;boolean> hasUnpartitionedCookieAccess();
+  Promise&lt;StorageAccessHandle> requestStorageAccess(StorageAccessTypes types);
+};
+</pre>
+
+When invoked on {{Document}} |doc|, the <dfn export method for=Document><code>hasUnpartitionedCookieAccess()</code></dfn> method must run these steps:
+
+1. Return the invocation of {{Document/hasStorageAccess()}}.
+
+Note:
+Now that {{Document/requestStorageAccess(types)}} <span class=allow-2119>can</span> be used to request [=unpartitioned data=] with or without specifically requesting cookies, it <span class=allow-2119>must</span> be made clear that {{Document/hasStorageAccess()}} only returns true if [=first-party-site context=] cookies are accessable to the current document.
+As a function name, {{Document/hasUnpartitionedCookieAccess()}} more clearly communicates this.
+For now {{Document/hasStorageAccess()}} is not considered deprecated, but that <span class=allow-2119>may</span> be worth taking up in future.
+
+When invoked on {{Document}} |doc|, the <dfn export method for=Document><code>requestStorageAccess(types)</code></dfn> method must run these steps:
+
+1. Let |p| be [=a new promise=].
+1. If |types|.{{StorageAccessTypes/all}} is `false` and |types|.{{StorageAccessTypes/cookies}} is `false`:
+    1. [=/Reject=] |p| with an "{{InvalidStateError}}" {{DOMException}}.
+    1. Return |p|.
+1. Let |requestUnpartitionedCookieAccess| be `true` if |types|.{{StorageAccessTypes/all}} is `true` or |types|.{{StorageAccessTypes/cookies}} is `true`, and `false` otherwise.
+1. Run the following steps [=in parallel=]:
+    1. Let |accessPromise| be the result of invoking {{Document/requestStorageAccess()}} with |requestUnpartitionedCookieAccess|.
+    1. If |accessPromise| [=/rejects=] with `reason` |r|:
+        1. [=/Reject=] |p| with |r|.
+    1. Else:
+        1. Let |handle| be a new object of type {{StorageAccessHandle}} with |types|.
+        1. [=/Resolve=] |p| with |handle|.
+1. Return |p|.
+
+
+<h3 id="request-storage-access-changes">Changes to {{Document/requestStorageAccess()}}</h3>
+
+Modify {{Document/requestStorageAccess()}} to invoke [=request storage access=] with |requestUnpartitionedCookieAccess| being `true`.
+
+Redefine the steps of {{Document/requestStorageAccess()}} to instead be the steps of <dfn export>request storage access</dfn> which takes a `boolean` argument |requestUnpartitionedCookieAccess|.
+
+Modify {{Document/requestStorageAccess()}} at step 14.1.1.1.1 to read:
+
+1. If |requestUnpartitionedCookieAccess| is `true`, then set <var ignore='monkeypatch'>global</var>'s [=environment/has storage access=] to true.
+
+Modify {{Document/requestStorageAccess()}} at step 14.3 to read:
+
+3. If |explicitSetting| is "`disallow`" and |requestUnpartitionedCookieAccess| is `true`:
```

Comment:
I don't get this one, why would we not reject access to (only) cross-site storage if cross-site cookies are disallowed? Maybe I'm misreading this line :)

Reply:
Maybe I misunderstood, if this just short circuits the cross-site cookies grant that's fine